lib/randomfileaccess.js

```diff
@@ -0,0 +1,167 @@
+    } catch (err) {
+        console.error(`Unable to close file handle of ${path}`, err);
+    }
+}
+
+/**
+ * Random file access 
+ * 
+ * This class holds the file handles for both "read" (upload) and "write" (download) so 
+ * independent {Transfer} function instances can read and write the same file concurrently.
+ */
+class RandomFileAccess {
+    /**
+     * Construct Random File Access
+     */
+    constructor() {
+        this[PRIVATE] = {
+            fileHandles: new Map()
+        };
+    }
+
+    /**
+     * Close the given file
+     * 
+     * If no path is provided, all files will be closed.
+     * 
+     * @param {String|URL} [path] Path to close
+     */
+    async close(path) {
+        if (path) {
+            const { fileHandles } = this[PRIVATE];
+            const fileHandlePromise = fileHandles.get(path);            
+            if (fileHandlePromise) {
+                fileHandles.delete(path);
+                await closeFile(path, fileHandlePromise);
+            }    
+        } else {
+            const closingFileHandles = this[PRIVATE].fileHandles;
+            this[PRIVATE].fileHandles = new Map();
+            for (const [path, fileHandlePromise] of closingFileHandles) {
+                await closeFile(path, fileHandlePromise);
+            }
+        }
+    }
+
+    /**
+     * Write a section to a file
+     * 
+     * @param {String|URL} path Path of the file to write to
+     * @param {SubRange} range Range of bytes to write
+     * @param {Buffer} buffer Buffer of bytes to write
+     * @param {Number} totalSize Total size of the file written
+     * @returns {WriteResponse} Bytes written and the buffer 
+     */
+    async write(path, range, buffer, totalSize) {
+        const handle = await this._getOrOpen(path, "w", totalSize);
+        return handle.write(buffer, 0, buffer.length, range.low);
+    }
+
+    /**
+     * Read a section of a file
+     * 
+     * @param {String|URL} path Path of the file to read from
+     * @param {SubRange} range Range of bytes to read
+     * @returns {Buffer} Buffer of bytes read
+     */
+    async read(path, range) {
+        const handle = await this._getOrOpen(path, "r");
+        const buffer = Buffer.allocUnsafe(range.length);
+        const { bytesRead } = await handle.read(buffer, 0, range.length, range.low);
+        return buffer.slice(0, bytesRead);
+    }
+
+    /**
+     * Get the file handle from cache, or open it if it's not already opened.
+     * 
+     * This is concurrency resilient.
+     * 
+     * @param {String|URL} path Path of file:// uri
+     * @param {String} flags Flags
+     * @param {Number} totalLength Total length of the file (write-only)
+     * @returns {FileHandle} File handle
+     */
+    async _getOrOpen(path, flags, totalLength) {
+        const { fileHandles } = this[PRIVATE];
+        
+        // store the promises so a concurrent read/write
+        // will end up with the same fileHandlePromise
```

Comment:
I fear this may lead to concurrency issues, though (especially if there are parallel writes).